### PR TITLE
Feature/storage market client

### PR DESCRIPTION
**Pull request by Alexey-N-Chernyshov**
### Description of the Change
Storage market client initial realization.

### CODE REVIEW

File: core/api/api.hpp [hunk 1]
```diff
@@ -154,6 +161,28 @@ namespace fc::api {
...
+    /**
+     * List client deals
+     */
+    API_METHOD(StateListClientDeals, std::vector<StorageDeal>, const Address &)
+
+    /**
+     * Adds funds with the StorageMinerActor for a storage participant. Used by
+     * both providers and clients.
+     */
+    API_METHOD(AddFunds, void, const Address &, const TokenAmount &)
```
Comment (turuslan) [inline core/api/api.hpp:180]: Remove `AddFunds`, call `MpoolPushMessage` and `StateWaitMsg`

File: core/api/api.hpp [hunk 2]
```diff
@@ -154,6 +161,28 @@ namespace fc::api {
...
+    API_METHOD(ValidateAskSignature,
+               bool,
+               const SignedStorageAsk &,
+               const TipsetKey &)
```
Comment (turuslan) [inline core/api/api.hpp:184]:
```suggestion
    API_METHOD(StateMinerInfo, MinerInfo, const Address &, const TipsetKey &)
```

File: core/api/api.hpp [hunk 3]
```diff
@@ -154,6 +161,28 @@ namespace fc::api {
...
+    API_METHOD(SignProposal, ClientDealProposal, Address, DealProposal)
```
Comment (turuslan) [inline core/api/api.hpp:185]:
```suggestion
    API_METHOD(StateAccountKey, Address, const Address &, const TipsetKey &)
```
Then sign using keystore

File: core/api/api.hpp [hunk 4]
```diff
@@ -154,6 +161,28 @@ namespace fc::api {
...
+    /**
+     * Lists the providers in the storage market state
+     * @return vector of StorageProviderInfo
+     */
+    API_METHOD(StateListStorageProviders, std::vector<StorageProviderInfo>)
+
+    /**
+     * List client deals
+     */
+    API_METHOD(StateListClientDeals, std::vector<StorageDeal>, const Address &)
```
Comment (turuslan) [inline core/api/api.hpp:174]: Remove `StateListClientDeals`, call `StateMarketDeals`

File: core/api/api.hpp [hunk 5]
```diff
@@ -154,6 +161,28 @@ namespace fc::api {
     API_METHOD(SyncSubmitBlock, void, const BlockMsg &)
 
     API_METHOD(WalletSign, Signature, const Address &, const Buffer &)
+
+    /**
+     * Lists the providers in the storage market state
+     * @return vector of StorageProviderInfo
+     */
+    API_METHOD(StateListStorageProviders, std::vector<StorageProviderInfo>)
```
Comment (turuslan) [inline core/api/api.hpp:169]:
```suggestion
    API_METHOD(StateListMiners, std::vector<Address>, const TipsetKey &)
```
Remove `StateListStorageProviders`, call `StateListMiners` and `StateMinerInfo `

File: core/markets/storage/ask_stream.hpp [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+#include "common/outcome.hpp"
+#include "markets/storage/ask_protocol.hpp"
+
+namespace fc::markets::storage {
+
+  /**
+   * StorageAskStream is a stream for reading/writing requests & responses on
+   * the Storage Ask protocol
+   */
+  class StorageAskStream {
```
Comment (turuslan) [inline core/markets/storage/ask_stream.hpp:18]: Remove deal and ask streams, use `CborStream`

File: core/markets/storage/client/client_impl.hpp [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    std::shared_ptr<StorageMarketNetwork> network_;
+    std::shared_ptr<data_transfer::Manager> data_transfer_manager_;
+    std::shared_ptr<IpfsDatastore> block_store_;
+    std::shared_ptr<FileStore> file_store_;
+    std::shared_ptr<PieceIO> piece_io_;
+
+    // TODO
+    // discovery
+
+    Api node;
```
Comment (turuslan) [inline core/markets/storage/client/client_impl.hpp:85]: Remove this

File: core/markets/storage/client/client_impl.cpp [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+                          const SignedAskHandler &signed_ask_handler) const {
+    network_->newAskStream(
+        info.peer_id,
+        [self{shared_from_this()}, info, signed_ask_handler](
+            auto &&stream_res) {
+          if (stream_res.has_error()) {
+            self->logger_->error("Cannot open stream to "
+                                 + info.peer_id.toBase58() + ": "
+                                 + stream_res.error().message());
+            return;
```
Comment (turuslan) [inline core/markets/storage/client/client_impl.cpp:67]: Should call callback?

File: core/api/api.hpp (hunk 1 above)
Comment (Alexey-N-Chernyshov) [reply core/api/api.hpp:180]: done

File: core/api/api.hpp (hunk 4 above)
Comment (Alexey-N-Chernyshov) [reply core/api/api.hpp:174]: done

File: core/api/api.hpp (hunk 5 above)
Comment (Alexey-N-Chernyshov) [reply core/api/api.hpp:169]: done

File: core/markets/storage/client/client_impl.cpp (hunk 8 above)
Comment (Alexey-N-Chernyshov) [reply core/markets/storage/client/client_impl.cpp:107]: done

**Pull request merged**